network/manager.go

```diff
@@ -0,0 +1,569 @@
+		if n := m.GetNodeInfo(msg.Src); n != nil {
+			resp.NodeInfo = append(resp.NodeInfo, n)
+		} else {
+			err = errNodeNotFound
+		}
+
+	case GET_NODES_I_ACCEPT:
+		list := m.GetAcceptNodes(msg.Src)
+		resp.NodeInfo = list
+
+	case GET_NODES_I_CAN_ACCESS:
+		list := m.GetNodesICanAccess(msg.Src)
+		resp.NodeInfo = list
+
+	case UPDATE_MY_INFO:
+		resp.NetworkInfo = m.networkData.NetworkInfo
+		err = m.UpdateName(msg.Src, req.Name)
+
+	default:
+		return nil, fmt.Errorf("nConnect manager got unknown message type: %v", req.MsgType)
+	}
+
+	if err != nil {
+		resp.Err = err
+	}
+
+	return resp, nil
+}
+
+func (m *Manager) JoinNetwork(address, name string) (*nodeInfo, error) {
+	if node, ok := m.networkData.Member[address]; ok {
+		node.LastSeen = time.Now()
+		m.networkData.Member[address] = node
+		if err := m.saveNetworkData(); err != nil {
+			return nil, err
+		}
+		noti := &managerToMember{
+			MsgType:  NOTI_MEMBER_ONLINE,
+			NodeInfo: []*nodeInfo{node},
+		}
+		SendMsg(m.c, node.Address, noti, false)
+
+		return node, nil
+	}
+
+	m.Lock()
+	defer m.Unlock()
+	_, ok := m.networkData.Waiting[address]
+	if ok {
+		return nil, errWaitForAuth
+	}
+
+	if _, nameExists := m.networkData.NameToAddress[name]; nameExists {
+		return nil, errNameExist
+	}
+
+	if len(name) > 0 {
+		m.networkData.NameToAddress[name] = address
+	}
+	m.networkData.Waiting[address] = &nodeInfo{Name: name, Address: address, LastSeen: time.Now()}
+
+	if err := m.saveNetworkData(); err != nil {
+		return nil, err
+	}
+
+	return nil, errWaitForAuth
+}
+
+func (m *Manager) LeaveNetwork(address, name string) error {
+	modified := false
+	if _, ok := m.networkData.Member[address]; ok {
+		delete(m.networkData.Member, address)
+		modified = true
+	}
+	m.Lock()
+	defer m.Unlock()
+	if _, ok := m.networkData.Waiting[address]; ok {
+		delete(m.networkData.Waiting, address)
+		modified = true
+	}
+	if modified {
+		return m.saveNetworkData()
+	}
+
+	return nil
+}
+
+func (m *Manager) AuthorizeMemeber(n *nodeInfo) error {
+	m.Lock()
+	defer m.Unlock()
+
+	if nw, ok := m.networkData.Waiting[n.Address]; ok {
+		ip, err := m.GetAvailableIp()
+		if err != nil {
+			log.Println("GetAvailableIp error:", err)
+			return err
+		}
+		nw.IP = ip
+		nw.Netmask = m.networkData.Netmask
+		m.networkData.Member[n.Address] = nw
+
+		if m.opts.Verbose {
+			log.Printf("AuthorizeMemeber, Member: %+v\n", m.networkData.Member)
+		}
+
+		delete(m.networkData.Waiting, n.Address)
+
+		if err = m.saveNetworkData(); err != nil {
+			return err
+		}
+
+		noti := &managerToMember{
+			MsgType:     NOTI_AUTHORIZED,
+			NetworkInfo: m.networkData.NetworkInfo,
+			NodeInfo:    []*nodeInfo{nw},
+		}
+
+		if _, err = SendMsg(m.c, n.Address, noti, false); err != nil {
+			log.Printf("Send NOTI_AUTHORIZED to %v error %v\n", n.Address, err)
+		}
+
+		for _, n := range m.networkData.Member {
+			acceptAddresses := m.networkData.AcceptAddress[n.Address]
+			if len(acceptAddresses) > 0 && acceptAddresses[0] == AcceptAll {

```

Comment:
Anything other than acceptall we should handle?

Reply:
The manager can set each member's accept nodes. One option is to accept all other members `AcceptAll`. other options the manager set only some nodes which can accept the designated node.